### PR TITLE
Replace `static mut`s with new `CoreLocal` construct

### LINKED ISSUE #3841: Shared references of mutable static warnings
While compiling tock with the latest nightly version of rust, `nightly-2024-02-08`, I encounter 25 warnings regarding the usage of shared references of mutable static in folders such as `arch/`, `kernel/`, `chips/` and `boards/` as it can be seen bellow:

```
warning: shared reference of mutable static is discouraged
   --> kernel/src/deferred_call.rs:145:28
    |
145 |         let ctr = unsafe { &CTR };
    |                            ^^^^ shared reference of mutable static
    |
    = note: for more information, see issue #114447 <https://github.com/rust-lang/rust/issues/114447>
    = note: reference of mutable static is a hard error from 2024 edition
    = note: mutable statics can be written to by multiple threads: aliasing violations or data races will cause undefined behavior
    = note: `#[warn(static_mut_ref)]` on by default
help: shared references are dangerous since if there's any kind of mutation of that static while the reference lives, that's UB; use `addr_of!` instead to create a raw pointer
    |
145 |         let ctr = unsafe { addr_of!(CTR) };
```

If I understand correctly, for the moment this is just a warning, but it will soon become a compile error. We will need some workarounds for this.

### PR DESCRIPTION
**Pull request by alevy**
### Pull Request Overview

This pull request fixes #3841 by replacing most global `static mut` throughout the kernel, chips, capsules, and board crates with a new construct called `CoreLocal`.

`CoreLocal` is similar in principle to Rust `std`'s `LocalKey` (obtained from the `thread_local!` macro). It allows access to its internally stored data within a closure passed to `with`, which obtains a temporary shared reference. It is marked `Sync`, under the assumption that it is accessed in a thread-safe way---e.g. one copy per CPU core, as the name implies---allowing it to be stored in a global (non-mut) static.

This implementation only includes support for single-core, single-threaded, platforms (as all of Tock for the moment) via the unsafe `new_single_core` constructor, though extending to multiple cores should be relatively simple.

In order to support mutation (which is ostensibly the purpose of `static mut` in many cases), users should use some form of single-threaded mutual exclusion, such as a `MapCell`, `TakeCell`, `Cell`, etc.

This PR converts most uses of `static mut` to use this new construct, thus avoiding various annoying warnings and _probably_ some real unsoundness. This comes at virtually no cost to code size, but some minor cost to memory (around 100 bytes on the platforms I looked at).

### Testing Strategy

Still need to test. This changes some subtle parts of the kernel (deferred calls, the schedulers, process implementation) in ways that, if I didn't take enough care to be precise about the logic, could likely result in bugs (though probably not safety bugs).

### TODO or Help Wanted

- [ ] Finish porting other boards. I _think_ this is just menial and straight forward, but tbd
- [ ] Port some chip-specific drivers that have more subtle uses of static mut

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

### CODE REVIEW

**Comment by alevy** [comment]
@lschuermann and I discussed this, so summarizing the feedback and take aways here:

* One potential issue is that we technically have additional threads in the ISRs.
* Those really really shouldn't touch the same `static mut`s (now `CoreLocal`s), but we don't have a great way of enforcing that
* This probably comes down to how we articulate safety for the constructor for `CoreLocal` or by adding `unsafe` to one or more methods.

Ideally, we need to do at least one of:

1. make it impossible or very hard to accidentally call the offending method from an ISR context, e.g., by limiting the state that ISRs can access (for instance, putting them in a separate crate, although that's not bulletproof)
2. make the constructor itself unsafe.
3. make the method that accesses it (`with`) unsafe.

The current implementation in this PR makes the constructor unsafe, but does not yet explain why it is unsafe, and what need to be ensured (e.g. that it is private to the module, and that it is only accessible by an ISR).

File: boards/nordic/nrf52840dk/src/main.rs [hunk 1]
```diff
@@ -914,8 +916,14 @@ pub unsafe fn start() -> (
     let _ = platform.pconsole.start();
     base_peripherals.adc.calibrate();
 
+    io::DEBUG_INFO.with(|debug_info| debug_info.put(io::DebugInfo {
+	chip,
+	processes,
+	process_printer,
+    }));
```
Comment (hudson-ayers) [inline boards/nordic/nrf52840dk/src/main.rs:923]: This passes references to these to a potential future panic "thread" while other references to them live on and continue to be used.

Do we have a good sense for whether this is safe? I agree that the old approach is probably unsafe, so this is not necessarily making things worse. But it is not obvious to me that this is safe, and if it is not I sort of think we shouldn't make it look like it may be.

Comment (lschuermann) [reply boards/nordic/nrf52840dk/src/main.rs:923]: The fundamental danger with this infrastructure is that we're providing a reference of `T: ?Sync` in a `Sync` wrapper, and I hardly think that `chip` or `processes` are `Sync`. So strictly speaking this is probably not sound as would be required by Rust.

This is a hard problem to solve, and while the introduction of `CoreLocal` does not change our premise and actual safety guarantees here, it's properties around `Sync` are what's problematic... In fact, I never thought of panics as being virtually a different thread (which they definitely are!), but only ISRs.

I think that this here is symptomatic of a larger body of unsoundness in Tock when it comes to reasoning about which state can be assumed to be valid in these non-single-thread-synchronous circumstances. A panic may occur precisely because an invariant in some struct reachable through those reference is violated, and Rust doesn't give us particularly many guarantees around their validity beyond that point. So, it's really not only that `chip` and `processes` are `!Sync`, but that they may not be valid at all in a panic context. An unergonomic, expensive and ugly fix is to move all state required in the panic handler into some "atomic" statics (perhaps wrapped in a `CoreLocal`), that are `Sync`, have a `Cell`-like interface, can only wrap datatypes that are guaranteed to be updated atomically w.r.t. a single hardware thread, and are unconditionally valid.

In short, I don't think we can or should fix this in this PR.

**Comment by lschuermann** [comment]
Agree with @alevy's summary, just to illustrate the point of why we reason about ISRs / panic contexts at all:

The fundamental assumption in this infrastructure is that, due to Tock's single-threaded nature, deriving `Sync` on a wrapper around a `T: ?Sync` is sound. However, we can violate this assumption through use of this infrastructure in ISR or (as pointed out by @hudson-ayers) panic contexts, which behave as virtually separate threads, bringing concerns about reentrancy with them. Even if we find this acceptable as it's not strictly worse than what we're doing right now anyways, it's still important that we strive to remove these soundness issues in the future. This is exactly where the proposals around either strictly separating out state that is accessible in these "threaded" contexts and/or making parts of this infrastructure unsafe, as summarized by @alevy above, come in.

File: kernel/src/core_local.rs
Comment (bradjc) [inline kernel/src/core_local.rs:1]: This should go in kernel/utilities.

File: kernel/src/kernel.rs [hunk 2]
```diff
@@ -41,10 +46,43 @@ use crate::utilities::cells::NumericCellExt;
 /// is less than this threshold.
 pub(crate) const MIN_QUANTA_THRESHOLD_US: u32 = 500;
 
+#[derive(Copy, Clone)]
+pub struct StaticSlice<T> {
```
Comment (bradjc) [inline kernel/src/kernel.rs:50]: ?

File: kernel/src/process_checker/basic.rs [hunk 3]
```diff
@@ -83,6 +83,7 @@ impl<'a> AppCredentialsPolicy<'a> for AppCheckerSimulated<'a> {
     }
 }
 
+#[allow(dead_code)]
```
Comment (bradjc) [inline kernel/src/process_checker/basic.rs:86]: ?

File: kernel/src/grant.rs [hunk 4]
```diff
@@ -1780,11 +1784,8 @@ pub struct Iter<
     /// The grant type to use.
     grant: &'a Grant<T, Upcalls, AllowROs, AllowRWs>,
 
-    /// Iterator over valid processes.
-    subiter: core::iter::FilterMap<
-        core::slice::Iter<'a, Option<&'static dyn Process>>,
-        fn(&Option<&'static dyn Process>) -> Option<&'static dyn Process>,
-    >,
+    processes: &'a CoreLocal<MapCell<StaticSlice<Option<&'static dyn Process>>>>,
+    iterator: usize,
```
Comment (bradjc) [inline kernel/src/grant.rs:1788]: We lose the comment. Why this change?

**Comment by bradjc** [comment]
> This implementation only includes support for single-core, single-threaded, platforms (as all of Tock for the moment) via the unsafe `new_single_core` constructor, though extending to multiple cores should be relatively simple.

At the moment seems to imply we need to guard against something else, but I don't see the need to make the scope larger than our current issue. Since this is pressing (we can't upgrade nightly), I think we need to keep the scope narrow.

> In order to support mutation (which is ostensibly the purpose of `static mut` in many cases), users should use some form of single-threaded mutual exclusion, such as a `MapCell`, `TakeCell`, `Cell`, etc.

This seems like something that should be wrapped in our type. Why put the complexity on to users? It seems like the _is_ the purpose of `static mut`, so why not a one-to-one replacement?


> * Those really really shouldn't touch the same `static mut`s (now `CoreLocal`s), but we don't have a great way of enforcing that
> 
> * This probably comes down to how we articulate safety for the constructor for `CoreLocal` or by adding `unsafe` to one or more methods.


I view our job as kernel developers to manage the complexity of using unsafe correctly. Exposing more APIs where the burden is on other developers to reason about `unsafe` correctness abdicates our responsibility. We should put safeguards in place to make writing correct code easier. If that means a CI script that checks ISR functions, then that is what is requires.

File: boards/nordic/nrf52840dk/src/main.rs (hunk 1 above)
Comment (alevy) [reply boards/nordic/nrf52840dk/src/main.rs:923]: Panic is not in a different thread

**Comment by alevy** [comment]
> This seems like something that should be wrapped in our type. Why put the complexity on to users? It seems like the is the purpose of static mut, so why not a one-to-one replacement?

Three reasons:

1. It separates reasoning about the safety of `CoreLocal` from reasoning about the safety of interior mutability. Both are a bit subtle, and to the extent they can be kept correct separately, the better.
2. Different users want/need different Cells for interior mutability. Already in the ports I have there are a couple cases where `Cell` makes more sense.
3. Mutability isn't the _only_ reason we use `static mut`. The other reason is that you're not allowed to put even read-only things in non-mut globals that don't implement `Sync`, which is most things. So we have a few places where we use `static mut` to store something that is never used as `mut`, but isn't `Sync`. `CoreLocal` without an internal cell is better (more efficient, easier to use, more obviously correct,  etc).

File: kernel/src/process_checker/basic.rs (hunk 3 above)
Comment (alevy) [reply kernel/src/process_checker/basic.rs:86]: It's a warning we get on newer compilers that for some reason maybe we weren't getting before? Either way, the code is never used and should probably actually be removed, but I didn't want to mess with that stuff in this PR. (Document in the changing commit, but I know that's somewhat hidden)

File: kernel/src/kernel.rs (hunk 2 above)
Comment (alevy) [reply kernel/src/kernel.rs:50]: A `[u8]` is not `Sized`, so can't be stored in a bunch of places. In Rust `std` statics (like `thread_local`) end up using `Box` or `Vec`. `StaticSlice` is basically the a fixed-sized `Vec` (it has a fixed length, but that length is stored dynamically), so it is `Sized` and can be stored in statics directly, like a fixed-sized array.

File: kernel/src/core_local.rs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+pub struct CoreLocal<T>(UnsafeCell<T>);
+
+impl<T> CoreLocal<T> {
+    pub const unsafe fn new_single_core(val: T) -> Self {
+	CoreLocal(UnsafeCell::new(val))
+    }
+}
+
+impl<T> CoreLocal<T> {
+    pub fn with<F, R>(&self, f: F) -> R where F: FnOnce(&T) -> R {
```
[the remark this answers is not in the document]
Comment (alevy) [reply kernel/src/core_local.rs:12]: Yes, it's how `thread_local` works in Rust's `std`

File: kernel/src/core_local.rs
Comment (alevy) [reply kernel/src/core_local.rs:1]: Oh, agree.

File: boards/nordic/nrf52840dk/src/main.rs (hunk 1 above)
Comment (alevy) [reply boards/nordic/nrf52840dk/src/main.rs:923]: > I hardly think that chip or processes are Sync

Why? Multiple points of entry does not mean they are multiple threads of execution. They are not multiple threads of execution and that is pretty important to the Tock kernel functioning correctly generally.

There are exactly two kinds of software things that run concurrently with the kernel:

- Actual Processes, which do not have access to kernel memory
- ISRs which are specifically designed to do almost nothing except switch back to the kernel thread and retain the interrupt flag. They really _should not_ touch any memory shared with the broader kernel.

Everything else is single threaded. If it is not single threaded, it is a bug.

**Comment by alevy** [comment]
Let me try to articulate the safety invariants for `CoreLocal` (that are not currently in the documentation, but obviously should be) that I think might make the feedback/discussion better directed.

A `CoreLocal` stores data (statically allocated) that is has a unique value per-core. Because Tock's execution model is such that each core has exactly one kernel-thread of execution, this struccture is `Sync` and, thus can wrap non-Sync types or references to non-Send types (most things in Tock). Meaning, it can be stored in non-mut statics.

## An aside for justification

This kind of co-reliance between the system's execution model and the language-enforced APIs is fundamental. The notion of a thread, what constitutes a thread, and how to distinguish different threads is entirely an OS-level construct. For example, the safety of `thread_local` in Rust std entirely relies on the particular implementation of pthreads (e.g. using the gs register to store a location to thread-local variables and the kernel swaps that register when switching threads).

## The hidden kernel threads

Tock _does_ actually have one other kind of software that runs truly-concurrently (i.e. it may run at any point, including during "critical sections") with the kernel---interrupt service routines (ISRs).

There's an important invariant for ISRs (which _may_ not always be maintained at the moment): ISRs must not share any data with the kernel _except_ the variables that signal a syscall, hardfault, etc just fired (even those are generally controlled _within_ the arch crate, but they are accessed in both ISR and kernel threads). Those do not use `CoreLocal`.

There are at least two notable exceptions to this rule right now. As @hudson-ayers points out, `panic` might be called from both the kernel thread and ISRs (for example, when an interrupt is not recognized, some archs `panic!`). This is a problem if the `panic` handler accesses core-local variables, which it does/will on most boards in order to print a debug message and blink.

## Supporting more than one thread-of-execution

Supporting more than one thread (e.g. ISRs) _safely_ is a tricky and potentially heavy-weight proposition. There needs to be _some_ way of determining whether we are currently in---e.g. are we in ISR or the kernel main-thread, and _which_ ISR. We _could_ do this in architecture specific ways---e.g. use the armv7m IPSR register (which denotes the exception level), and either deny access if it's not the kernel main-thread, or have a separate copy for each ISR (that's a lot of memory....).

This implementation argues that a better strategy is to eliminate access to CoreLocals from ISRs, use an unsafe constructor for CoreLocal to denote that the user must ensure that such a static is not accessible from another core or from an ISR, and to have a simple set of "rules" to follow that ensure this by convention---and rely on ISRs themselves being highly subtle from a safety perspective (and requiring unsafe to implement in the first place) to avoid violating these rules in ISR context.

File: kernel/src/process_checker/basic.rs (hunk 3 above)
Comment (bradjc) [reply kernel/src/process_checker/basic.rs:86]: I think it should be `pub`.

File: boards/nordic/nrf52840dk/src/main.rs (hunk 1 above)
Comment (lschuermann) [reply boards/nordic/nrf52840dk/src/main.rs:923]: > I hardly think that chip or processes are Sync

Aside from your actual comment text which I agree with, the aforementioned variables `chip` and `process` won't be sync because at least, as far as I know, they all contain some form of `core::cell::UnsafeCell` that is `!Sync`.

Comment (alevy) [reply boards/nordic/nrf52840dk/src/main.rs:923]: Oh... I retract! You're of course right. I totally misunderstood that you were referring to the variables rather than the crates/modules. Sorry!

**Comment by bradjc** [comment]
> This is a problem if the `panic` handler accesses core-local variables, which it does/will on most boards in order to print a debug message and blink.

I don't think this is problem worth thinking about; if your ship's sinking, poking a hole in the hull isn't going to matter.


> This implementation argues that a better strategy is to eliminate access to CoreLocals from ISRs,

Agreed.

> use an unsafe constructor for CoreLocal to denote that the user must ensure that such a static is not accessible from another core

This seems like it needs more explanation as to what the expectation is.

> or from an ISR,

Agreed.

> and to have a simple set of "rules" to follow that ensure this by convention

Sounds good in theory, but what are they?

---and rely on ISRs themselves being highly subtle from a safety perspective (and requiring unsafe to implement in the first place) to avoid violating these rules in ISR context.

> Very much agreed.

File: kernel/src/kernel.rs [hunk 6]
```diff
@@ -229,18 +257,22 @@ impl Kernel {
...
-                    if ret.is_some() {
-                        return ret;
-                    }
-                }
-                None => {}
-            }
-        }
-        None
+	self.processes.with(|ps| {
+	    ps.map_or(None, |ps| {
```
Comment (bradjc) [inline kernel/src/kernel.rs:261]: Does `if let Some(ps) = ps.get()` not work here?

File: kernel/src/kernel.rs [hunk 7]
```diff
@@ -41,10 +46,43 @@ use crate::utilities::cells::NumericCellExt;
...
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe { core::slice::from_raw_parts_mut(self.base, self.len) }
+    }
+}
+
 /// Main object for the kernel. Each board will need to create one.
 pub struct Kernel {
     /// This holds a pointer to the static array of Process pointers.
-    processes: &'static [Option<&'static dyn process::Process>],
+    pub(crate) processes: &'static CoreLocal<MapCell<StaticSlice<Option<&'static dyn process::Process>>>>,
```
Comment (bradjc) [inline kernel/src/kernel.rs:85]: Why `pub`?

Comment (bradjc) [inline kernel/src/kernel.rs:85]: I'm concerned about adding `MapCell` here. Nothing should ever call `.take()`, but that is an option now. What we need is a cell that can't lose its contents.

**Review by bradjc — COMMENTED** [review]
There's a lot in this PR (two new kernel types, changes to HILs, wrapping static muts in cells, and of course the headline change). How should we make forward progress?

File: kernel/src/kernel.rs [hunk 8]
```diff
@@ -41,10 +46,43 @@ use crate::utilities::cells::NumericCellExt;
...
+    fn deref(&self) -> &Self::Target {
+        unsafe { core::slice::from_raw_parts(self.base, self.len) }
+    }
+}
+
+impl<T> DerefMut for StaticSlice<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe { core::slice::from_raw_parts_mut(self.base, self.len) }
+    }
+}
```
Comment (jrvanwhy) [inline kernel/src/kernel.rs:80]: `StaticSlice` is unsound. Because `StaticSlice` is `Copy`, you create a `StaticSlice` pointing to a buffer, duplicate it, then call `deref_mut` on both copies to get two `&mut` references to the same buffer.

Demonstration: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=21831ccf86ae6ebcc70ef9d9263a9fb4

File: kernel/src/kernel.rs (hunk 2 above)
Comment (jrvanwhy) [reply kernel/src/kernel.rs:50]: I'm having trouble seeing the semantic difference between a `StaticSlice<T>` and a `&'static mut [T]`.

File: chips/nrf52/src/ble_radio.rs [hunk 9]
```diff
@@ -592,9 +588,21 @@ impl<'a> Radio<'a> {
...
+            let old_base = self.registers.packetptr.get() as usize as *mut u8;
+	    let old_len = if old_base.is_null() {
+		0
+	    } else {
+		self.registers.pcnf1.read(PacketConfiguration1::MAXLEN) as usize
+	    };
+            self.registers.packetptr.set(buf.as_ptr() as u32);
+	    self.registers.pcnf1.modify(PacketConfiguration1::MAXLEN.val(buf.len() as u32));
+	    slice::from_raw_parts_mut(old_base, old_len)
         }
```
Comment (jrvanwhy) [inline chips/nrf52/src/ble_radio.rs:606]: It looks like `old_base` can be null at the `slice::from_raw_parts_mut` invocation, which is UB.

Please add a safety comment for `slice::from_raw_parts_mut`, and minimize the size of this `unsafe` block (is `slice::from_raw_parts_mut` the only unsafe operation here?).

File: kernel/src/core_local.rs [hunk 10]
```diff
@@ -0,0 +1,17 @@
+use core::cell::UnsafeCell;
+
+pub struct CoreLocal<T>(UnsafeCell<T>);
```
Comment (jrvanwhy) [inline kernel/src/core_local.rs:3]: I don't see why we need `UnsafeCell` here, though maybe the documentation will clarify that when it is added.

**Comment by jrvanwhy** [comment]
> Need documentation.

In particular, I'd like to see safety requirements of `CoreLocal::new_single_core` -- it's almost impossible to review the rest of this PR for soundness without that documentation.

File: kernel/src/core_local.rs [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+impl<T> CoreLocal<T> {
+    pub const unsafe fn new_single_core(val: T) -> Self {
+	CoreLocal(UnsafeCell::new(val))
+    }
+}
```
[the remark this answers is not in the document]
Comment (jrvanwhy) [reply kernel/src/core_local.rs:9]: That's not how I read it. A `new_<something>()` in Rust generally describes creating a particular variant of a thing (in this case, creating a single-core variant of a `CoreLocal`).

**Comment by ppannuto** [comment]
Catching up on phone on the train ride home, so haven't read or thought through all the subtleties here, but I feel it's worth pointing out that ISRs need not **necessarily** risk running concurrently with accesses to CoreLocals; i.e., it is a single assembly instruction to turn all interrupts off and similarly one to reenable them. We have the fundamental primitive available if needed to meet Rust's safety requirements IIUC. 

(There are further complexities to such an approach, and overhead that isn't strictly necessary if we can otherwise prove ISRs won't access the memory, but a "Rust-legal Sync" is somewhat trivially available from this design IIUC)

File: kernel/src/core_local.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (bradjc) [reply kernel/src/core_local.rs:9]: I agree that new_X is a variant, but it's almost always an adjective (eg unchecked) not a noun (core).

https://doc.rust-lang.org/std/?search=new_

There seem to be exceptions for creating a slice.

Maybe this should be `new_single`? Although I don't see the need for the modifier if there is only one new() method.

File: kernel/src/core_local.rs (hunk 10 above)
Comment (lschuermann) [reply kernel/src/core_local.rs:3]: We don't, I believe! I'm confident this is an artifact from our earlier design. We may want to re-introduce it in the future, if we ever add an `unsafe fn with_mut`, where developers themselves must ensure non-reentrance.

File: kernel/src/kernel.rs (hunk 2 above)
Comment (alevy) [reply kernel/src/kernel.rs:49]: There should be no semantic difference, only that `&'static mut [T]` is not `Sized` so cannot be declared in a context that needs `Sized` (such as a a static variable).

File: kernel/src/kernel.rs (hunk 8 above)
Comment (alevy) [reply kernel/src/kernel.rs:79]: This was an oversight when I was just trying to get the uses of it elsewhere in kernel to work. I've fixed this, it is no longer `Copy` or `Clone` and, you're right, absolutely should not be.

File: kernel/src/core_local.rs [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    ///
+    /// Even on single-core systems with no thread runtime, there may
+    /// be additional threads of execution, such as Interrupt Service
+    /// Routines (ISRs) or signal handlers, that could race with the
+    /// main thread if sharing a [`CoreLocal`]. It is safety-critical
+    /// that such contexts do not access [`CoreLocal`]s.
+    ///
+    /// By convetion, users should declare [`CoreLocal`] variables in
+    /// scopes that are inaccessible to ISRs or signal handler,
+    /// e.g. in module-private or function-local scopes.
```
Comment (jrvanwhy) [inline kernel/src/core_local.rs:63]: This comment is somewhat self-contradictory. It first says "must only be created ... on systems ... without preemptive threads". It then explains that single-core systems can have preemptive threads in the form of ISRs and signal handlers, and implies it is okay to use a `CoreLocal` on such systems as long as you hide the `CoreLocal` from ISRs and signal handlers.

File: kernel/src/kernel.rs (hunk 2 above)
Comment (jrvanwhy) [reply kernel/src/kernel.rs:49]: Ah, I see. The issue is not `Sized` but that `mut` references are not allowed in a `static` item's type. Maybe a different name and/or a link to https://github.com/rust-lang/rust/issues/57349 would make this more clear?

Some other improvement ideas:

1. `StaticSlice` doesn't actually need to be slice-specific. You can make it generic for any type and let the user choose whether to store a slice in it.
2. You can store a `&'static mut T` instead of a raw pointer to avoid `unsafe` in `Deref` and `DerefMut`.

Demonstration: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8de22b70edcbedaffccc382fd35ddd11

**Comment by bradjc** [comment]
I'm trying to understand the balance of flexibility with specificity of `CoreLocal`. The naming suggests flexibility: these objects are local to a core on CPUs with a single core, implying that this also supports objects local to a core on a multi-core CPU, and that another hypothetical type supports objects global to cores on multi-core CPUs. Given the scope of Tock (single core MCUs), I don't see the need for defining this broad space of shared objects. At the very least, the existence of `new_single_core()` implies there should be a `new_multi_core()`, otherwise, why be so specific? If the type only works on single core, then it would make sense for the type to be `SingleCoreLocal` with just `new()`.

If there is a clear need for defining types for a more complex multi-threaded platform, what does the upgrade path look like for users who want to use the hypothetical multicore and/or multithread versions? If the kernel uses `CoreLocal::new_single_core()` everywhere, should they run a local patch to replace those types and constructors with the multiX version?

It seems like the current CoreLocal gets us to an unclear middleground: the type & constructor imply multiX support based on their naming, but don't indicate how you would use them to run on multiX systems.

-----

One idea is to name the type based on how it should be used, not how it is implemented. Maybe `SharedGlobal`. The implementation changes based on the system you are running on to provide safety on that system. (Given there is no ambiguity about the hardware support for a given a system, this configuration is concrete and clear.) However, we want an implementation right now which does not make using this type fully safe in any context. So this may not be the approach right now.

The other direction is to double down on specificity and implementation, and not plan for a more generic approach right now. Maybe `MainThreadSingleCoreLocal` and then just a `::new()` constructor. This, however, may also imply the existence of `AnyThreadSingleCoreLocal` and `MultiCoreLocal` etc. types. I think a more custom name (eg `SimpleGlobal`) would be fine too.

**Comment by bradjc** [comment]
Are we still pursuing this?

**Comment by jrvanwhy** [comment]
I recently stumbled across a [Zulip discussion from November](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/More.20general.20term.20for.20.22threads.22) about `Send` and `Sync` for threads and thread-like things. I don't think the conversation conflicts with anything in this PR, but I think it's good to understand how other Rust users (including the rust-embedded WG, language team, and opsem team) handle these problems.
